tensorboard/uploader/proto/write_service.proto

```diff
@@ -119,6 +121,82 @@ message WriteScalarResponse {
   // This is empty on purpose.
 }
 
+// The blob writing procedure works as follows:
+//  * The caller instantiates a WriteBlob RPC, which is a bidirectional stream.
+//  * The caller sends a number of WriteBlobRequests, providing the blob data
+//    in an ordered sequence of chunks.
+//  * The request size should be less than 4 MiB (2^20 = 4,194,304), because
+//    that is the default request limit for gRPC.  Accounting for the other
+//    fields of the request, then, and leaving some padding for safety, the data
+//    chunk should be somewhat smaller-- perhaps simply 4e6 bytes.
+//  * The first of these requests must provide write_blob_first_request_data,
+//    and the subsequent ones must not.
+//  * The last of these requests must set finalize_object=True, and the prior
+//    ones must not.
+//  * For a blob <= 4MB, a single chunk suffices, so the single request is both
+//    the first and the last.
+//  * The caller receives a sequence of WriteBlobResponses in return.  Each
+//    WriteBlobResponse corresponds to one WriteBlobRequest in order, reporting
+//    the state of the write operation.
+//  * The caller may choose to verify each response before sending the next
+//    request, or it may continue sending requests optimistically.
+//
+// Note that the WriteBlobRequest does not mirror the nested structure of
+// WriteScalarRequest, because we only ever send one blob at a time.
+
+// Metadata included with only the first chunk of the blob write stream.
+message WriteBlobFirstRequestData {
+  string experiment_id = 1;
+  string run = 2;
+  string tag = 3;
+  int64 step = 4;
+  int64 index = 5;
```

Comment:
Does this mean that it’s generally possible to append to the blob
sequence at a particular step? I had assumed that the client would be
required to specify the sequence length atomically at initial write
time, because conceptually the whole sequence is one unit. (Just as it
would be weird to show the user a torn state where we’ve received only
the first 4 bytes of their scalar, it seems slightly strange to show a
torn state where we’ve received only the first blob in a sequence.
Recall that sequences aren’t homogeneous—for instance, images are
represented as `[width, height, ...pngs]`.)

Relatedly, when we want to support resumable uploads, how will we know
whether there remain blobs that need to be appended to any sequence?

Reply:
Refactoring in the works, pending offline discussion.

Reply:
I think the updated design re CreateBlobSequence vs WriteBlob fully clarifies this.  Basically I now allow writing the blobs in any order, and in a resumable manner.

Re the corresponding internal change, I didn't add an object_state to the BlobSequence table, because it's easy enough to select all of the blobs and check their object_states.  Indeed that will happen naturally on the read path anyway, so we can handle the incomplete-upload scenario appropriately there.